atomic_reactor/plugins/pre_fetch_maven_artifacts.py

```diff
@@ -0,0 +1,216 @@
+
+        nvr_requests = util.read_yaml(file_path, default=[],
+                                      schema='schemas/fetch-artifacts-nvr.json')
+        return [NvrRequest(**nvr_request) for nvr_request in nvr_requests]
+
+    def read_url_requests(self):
+        file_path = os.path.join(self.workdir, self.URL_REQUESTS_FILENAME)
+        if not os.path.exists(file_path):
+            self.log.debug('%s not found', self.URL_REQUESTS_FILENAME)
+            return []
+
+        return util.read_yaml(file_path, default=[],
+                              schema='schemas/fetch-artifacts-url.json')
+
+    def process_by_nvr(self, nvr_requests):
+        download_queue = []
+
+        for nvr_request in nvr_requests:
+            build_info = self.session.getBuild(nvr_request.nvr)
+            if not build_info:
+                raise ValueError('Build {} not found.'.format(nvr_request.nvr))
+
+            maven_build_path = self.path_info.mavenbuild(build_info)
+            build_archives = self.session.listArchives(buildID=build_info['id'],
+                                                       type='maven')
+            build_archives = nvr_request.match_all(build_archives)
+
+            for build_archive in build_archives:
+                maven_file_path = self.path_info.mavenfile(build_archive)
+                # NOTE: Don't use urljoin here because maven_build_path does
+                # not contain a traling slash, which causes the last dir to
+                # be dropped.
+                url = maven_build_path + '/' + maven_file_path
+                download_queue.append(DownloadRequest(url, maven_file_path,
+                                                      build_archive['checksum']))
+
+            unmatched_archive_requests = nvr_request.unmatched()
+            if unmatched_archive_requests:
+                raise RuntimeError('NVR request for "{}", failed to find archives for: "{}"'
+                                   .format(nvr_request.nvr, unmatched_archive_requests))
+
+        return download_queue
+
+    def process_by_url(self, url_requests):
+        download_queue = []
+
+        for url_request in url_requests:
+            url = url_request['url']
+
+            if self.allowed_domains:
+                netloc = urlparse(url).netloc.lower()
+                if netloc not in self.allowed_domains:
+                    raise ValueError('Domain {} is not in list of allowed domains: {}'
+                                     .format(netloc, self.allowed_domains))
+
+            checksum = url_request['md5sum']
+            target = url_request.get('target', url.rsplit('/', 1)[-1])
+            download_queue.append(DownloadRequest(url, target, checksum))
+
+        return download_queue
+
+    def download_files(self, downloads):
+        artifacts_path = os.path.join(self.workdir, self.DOWNLOAD_DIR)
+
+        self.log.debug('%d files to download', len(downloads))
+
+        for index, download in enumerate(downloads):
```

Comment:
well it depends from where it will be downloading, but unless each source has upload speed limit, we are limited just by our download speed, which isn't as well limited per download, so one-by one is fine without any limitations